### PR TITLE
Mm 49227

### PR DESCRIPTION
**Pull request by neallred**
#### Summary
Add e2e tests for self-hosted purchase flow. One step in particular, setting the license public key, I don't think we have a good way to automate (yet), so for now this is something that needs to be run locally.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-49227

#### Screenshots
Successful local run of tests, following the setup steps described in comments in the file:

https://user-images.githubusercontent.com/13738432/211114586-3291ae7f-6e45-4b99-abd3-a20f7e6dbeec.mp4

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts
Comment (yasserfaraazkhan) [inline e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts:13]: can we add @self-hosted tag?

I need your input here. All the jobs in cypress-automation currently running have [license set up ](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/blob/e2e-master-prod/.gitlab-ci.yml#L124)

to run this in CI, should we add condition to skip license?
I am aiming to run this and other self-hosted test

File: e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts
Comment (yasserfaraazkhan) [inline e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts:146]: @neallred do you know if there's an api that we can block/alter to make it seem as 'instance' without license.
If we eventually, run this along with other tests, it permanently remove license in CI

File: e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts
Comment (yasserfaraazkhan) [inline e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts:144]: MM-T5367

File: e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts
Comment (yasserfaraazkhan) [inline e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts:329]: MM-T5369

File: e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts
Comment (yasserfaraazkhan) [inline e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts:286]: MM-T5368

File: e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts
Comment (neallred) [reply e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts:248]: We could, but it already is an assertion. If the content isn't there, the test will fail with `Timed out retrying after 20000ms: Expected to find content: 'Paid' but never did.` We don't perform any sort of visible/non visible or opacity operations on the text, so I don't see any added value of additional assertions like `.should('be.visible');` other than a lot of existing code does that. And we're already asserting on a lot of other more specific things that already give us a high level of confidence.

File: e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts
Comment (neallred) [reply e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts:146]: I think we could fake it for this test, but that we wouldn't want to because even if we did, this test still alters the license from what other tests expect (professional plan is purchased and applied, but other tests assume an enterprise level license). Not sure how doable this is with the existing test framework, but I think a more robust approach would be that after this test runs, it restores the license that all other tests depend on. Running this test automatically is also blocked by  https://mattermost.atlassian.net/browse/MM-49161 .

File: e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts
Comment (neallred) [reply e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts:13]: I think we can add any tags we want, but it looks to me that nothing is done automatically with them unless we configure it in the cypress-ui-automation repo: https://git.internal.mattermost.com/qa/cypress-ui-automation/-/blob/e2e-master-prod/.gitlab-ci.yml#L44 .

I think `@self-hosted` is too broad of a tag. Maybe we could call it `@hosted_customer` instead to narrow the focus to parts of the app related to license purchase? One way I could see us using `@hosted_customer` is to add it on to the `--sortLast` group, even after `@license_removal` group. That way, it wouldn't interfere with other tests. Does that usage make sense to you. Are there other usages you are thinking of?

**Comment by neallred** [comment]
@yasserfaraazkhan anything else on this one, or does it look ok now? I added a hosted_customer tag.

**Review by yasserfaraazkhan — APPROVED** [review]
Thank you @neallred .
I lost track of this. But changes look great.
once merged I will try running this on CI

**Comment by neallred** [comment]
@marianunez gentle ping for review

File: e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts
Comment (marianunez) [inline e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts:28]: Is this duplication check intended?

File: e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts
Comment (marianunez) [inline e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts:12]: Should this be `@prod` if we can only run this locally at the time?

File: e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts
Comment (marianunez) [inline e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts:315]: Curious, where do we add additional users besides the admin one?

File: e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts
Comment (marianunez) [inline e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts:318]: Should we check that it contains the error message in this case?

**Review by marianunez — APPROVED** [review]
Apologies for the delay on this. Left some non-blocking comments below. 

Mostly, wondering, what are the plans to integrate this into the `prod` execution runs? We may need to revisit the configuration of a CWS for self-hosted servers ?

File: e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts
Comment (neallred) [reply e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts:146]: Agree that that is ideal, but the current test run setup makes assumptions on that license and I think lacks the setup to make sure the right license is applied. So I think it makes sense to sort this to the end of the run group: https://git.internal.mattermost.com/qa/cypress-ui-automation/-/blob/e2e-master-prod/.gitlab-ci.yml#L44.

File: e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts
Comment (neallred) [reply e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts:12]: Good catch! It should be `@dev` for now.

File: e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts
Comment (neallred) [reply e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts:28]: nope

File: e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts
Comment (neallred) [reply e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts:184]: Yes, for up to 2 minutes. export compliance screen can take a long time, up to a minute, and in practice for dev screening I've seen both successes close to this limit and timeouts after this limit.

File: e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts
Comment (neallred) [reply e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts:314]: Lots of tests add a new user as they go along, and in local runs we don't know how many users there will already be. The `apiInitSetup` helper is probably the biggest creator of users in e2e tests: https://github.com/mattermost/mattermost-webapp/blob/master/e2e/cypress/tests/support/api/setup.ts#L22-L35

**Comment by neallred** [comment]
> Mostly, wondering, what are the plans to integrate this into the prod CI execution runs? We may need to revisit the configuration of a CWS for self-hosted servers ?

I think integrating this into prod CI runs is dependent on the implementation of the story to use test license public key https://mattermost.atlassian.net/browse/MM-49161 . After that, as long as the runs point to RFQA CWS and we set EnableDeveloper to false for this test, I think the test will work.

**Pull request merged**
